software/hdmi2pcie/kernel/main.c

```diff
@@ -0,0 +1,744 @@
+	vdev->release = video_device_release_empty;
+
+	vdev->fops = &hdmi2pcie_fops;
+	vdev->device_caps = V4L2_CAP_READWRITE | V4L2_CAP_STREAMING;
+
+	if (dir == HDMI2PCIE_DIR_IN) {
+		vdev->ioctl_ops = &hdmi2pcie_ioctl_in_ops;
+		vdev->device_caps |= V4L2_CAP_VIDEO_CAPTURE;
+	} else {
+		vdev->ioctl_ops = &hdmi2pcie_ioctl_out_ops;
+		vdev->device_caps |= V4L2_CAP_VIDEO_OUTPUT;
+	}	
+
+
+	vdev->lock = &chan->lock;
+	vdev->queue = q;
+	vdev->vfl_dir = dir ? VFL_DIR_TX : VFL_DIR_RX;
+	vdev->v4l2_dev = &chan->v4l2_dev;
+	video_set_drvdata(vdev, chan);
+
+	ret = video_register_device(vdev, VFL_TYPE_GRABBER, -1);
+	if (ret) {
+		dev_err(&pdev->dev, "video_register_device failed, ret=%d\n", ret);
+		goto v4l2_unregister;
+	}
+
+	chan->timings = timings_def;
+
+	dev_info(&pdev->dev, "V4L2 HDMI2PCIe driver loaded");
+	return 0;
+
+v4l2_unregister:
+	v4l2_device_unregister(&chan->v4l2_dev);
+	return ret;
+}
+
+static int hdmi2pcie_pci_probe(struct pci_dev *pdev, const struct pci_device_id *id)
+{
+	int i;
+	int ret;
+	uint8_t rev_id;
+
+	struct device *dev = &pdev->dev;
+	struct hdmi2pcie_priv_data *priv;
+
+	dev_info(dev, "\e[1m[Probing device]\e[0m\n");
+
+	priv = kzalloc(sizeof(*priv), GFP_KERNEL);
+	priv->chan_cnt = 2;
+	if(!priv) {
+		dev_err(dev, "Cannot allocate memory\n");
+		ret = -ENOMEM;
+		goto fail1;
+	}
+	priv->channels = kzalloc(sizeof(*priv->channels) * priv->chan_cnt, GFP_KERNEL);
+	if(!priv->channels) {
+		dev_err(dev, "Cannot allocate memory\n");
+		ret = -ENOMEM;
+		goto fail1;
+	}
+
+	for (i = 0; i < priv->chan_cnt; i++)
+		priv->channels[i].common = priv;
+
+	pci_set_drvdata(pdev, priv);
+	priv->pdev = pdev;
+
+	ret = pci_enable_device(pdev);
+	if (ret != 0) {
+		dev_err(dev, "Cannot enable device\n");
+		goto fail1;
+	}
+
+	/* check device version */
+	pci_read_config_byte(pdev, PCI_REVISION_ID, &rev_id);
+	if (rev_id != 1) {
+		dev_err(dev, "Unsupported device version %d\n", rev_id);
+		goto fail2;
+	}
+
+	if (pci_request_regions(pdev, HDMI2PCIE_NAME) < 0) {
+		dev_err(dev, "Could not request regions\n");
+		goto fail2;
+	}
+
+	/* check bar0 config */
+	if (!(pci_resource_flags(pdev, 0) & IORESOURCE_MEM)) {
+		dev_err(dev, "Invalid BAR0 configuration\n");
+		goto fail3;
+	}
+
+	priv->bar0_addr = pci_ioremap_bar(pdev, 0);
+	priv->bar0_size = pci_resource_len(pdev, 0);
+	priv->bar0_phys_addr = pci_resource_start(pdev, 0);
+	if (!priv->bar0_addr) {
+		dev_err(dev, "Could not map BAR0\n");
+		goto fail3;
+	}
+
+	pci_set_master(pdev);
+	ret = pci_set_dma_mask(pdev, DMA_BIT_MASK(32));
+	if (ret) {
+		dev_err(dev, "Failed to set DMA mask\n");
+		goto fail4;
+	};
+
+	ret = pci_enable_msi(pdev);
+	if (ret) {
+		dev_err(dev, "Failed to enable MSI\n");
+		goto fail4;
+	}
+
+	ret = hdmi2pcie_register_video_dev(pdev, &priv->channels[0], HDMI2PCIE_DIR_IN);
+	ret += hdmi2pcie_register_video_dev(pdev, &priv->channels[1], HDMI2PCIE_DIR_OUT);
+	if (ret) {
+		dev_err(dev, "Failed to register V4L2 device");
+		goto fail5;
+	}
+
+	return 0;
+
+fail5:
```

Comment:
Can you please use better names than `fail5` -- something like `fail_before_register_video_dev` might be better...

Reply:
`goto` labels should now use more meaningful names